### PR TITLE
http-gateway: Implement Web of Things Interface

### PR DESCRIPTION
**Pull request by Danielius1922**
## New Features
- **/things Endpoint**: Introduced a new endpoint, `/things`, to retrieve resource links for devices. This endpoint provides a standardized way to access and interact with connected devices.
  
## Tests
- **Endpoint Testing**: Added comprehensive tests for the new `/things` endpoint. These tests ensure that device resources are properly retrieved and formatted, maintaining robustness and reliability in various scenarios.

## Refactor
- **JSON Unmarshalling**: Enhanced the JSON unmarshalling process in HTTP gateway tests. This improvement focuses on better error handling and more efficient processing of JSON data, contributing to overall test stability and readability.

## Documentation
- **Updated Documentation**: Revised the documentation to reflect changes in the usage of constants and parameters during virtual device creation. The updates include support for conditional resource addition based on the availability of device descriptions, making the documentation more accurate and helpful.

## New Endpoints
- **/api/v1/things**: Implemented the `/api/v1/things` endpoint to list all device resources.
- **/api/v1/things/{deviceID}**: Implemented the `/api/v1/things/{deviceID}` endpoint to access resources specific to a device identified by `deviceID`.

## Miscellaneous
- **Golang Upgrade**: Upgraded the Golang version to 1.22, ensuring compatibility with the latest features and improvements in the Go ecosystem.

### CODE REVIEW

**Comment by Danielius1922** [comment]
@coderabbitai review

**Comment by Danielius1922** [comment]
@coderabbitai review

File: Makefile [hunk 1]
```diff
@@ -225,7 +225,67 @@ simulators: simulators/clean
 	$(call RUN-DOCKER-DEVICE,$(DEVICE_SIMULATOR_RES_OBSERVABLE_NAME),$(DEVICE_SIMULATOR_RES_OBSERVABLE_IMG))
 .PHONY: simulators
 
-env/test/mem: clean certificates nats mongo privateKeys scylla
+BRIDGE_DEVICE_SRC_DIR = $(WORKING_DIRECTORY)/test/bridge-device
+# BRIDGE_DEVICE_IMAGE = ghcr.io/plgd-dev/device/bridge-device:vnext
+BRIDGE_DEVICE_IMAGE = ghcr.io/plgd-dev/device/bridge-device:vnext-pr464
```
Comment (Danielius1922) [inline Makefile]: TODO: restore vnext tag before merge

File: test/test.go [hunk 2]
```diff
@@ -824,33 +830,41 @@ func MustGetHostname() string {
...
 }
 
-func MustFindDeviceByName(name string) (deviceID string) {
+func MustFindDeviceByName(name string, getResourceOpts ...device.GetResourceOpts) (deviceID string) {
 	var err error
 	for i := 0; i < 3; i++ {
 		ctx, cancel := context.WithTimeout(context.Background(), 3*time.Second)
 		defer cancel()
-		deviceID, err = device.FindDeviceByName(ctx, name, nil)
+		deviceID, err = device.FindDeviceByName(ctx, name, getResourceOpts...)
```
[the remark this answers is not in the document]
Comment (jkralik) [reply test/test.go:838]: If function have prefix `Must` in golang, the error case is handled by panic.

File: http-gateway/service/getThings.go [hunk 3]
```diff
@@ -0,0 +1,387 @@
...
+	}
+	var td wotTD.ThingDescription
+	ThingSetSecurity(&td, requestHandler.openIDConfig)
+
+	things := GetThingsResponse{
+		Base:                requestHandler.config.UI.WebConfiguration.HTTPGatewayAddress + uri.Things,
+		Links:               links,
+		Security:            td.Security,
+		SecurityDefinitions: td.SecurityDefinitions,
+		ID:                  "urn:uuid:" + hubCfg.GetId(),
```
Comment (Danielius1922) [inline http-gateway/service/getThings.go:101]: bridgeDeviceTD.GetThingDescriptionID(hubCfg.GetId())

File: http-gateway/service/getThings.go [hunk 4]
```diff
@@ -0,0 +1,387 @@
...
+		}
+		_, ok := validateProperties[href]
+		if !ok {
+			_, ok = validateProperties["/"+href]
+		}
+		if !ok {
+			delete(td.Properties, href)
+			continue
+		}
+		patchedProp, err := patchProperty(prop, deviceID, href, message.AppJSON)
```
Comment (Danielius1922) [inline http-gateway/service/getThings.go:288]: do we want to hard code json format?

Comment (jkralik) [reply http-gateway/service/getThings.go:288]: for now yes, in future we can process accept header.

File: http-gateway/service/getThings.go (hunk 3 above)
Comment (jkralik) [reply http-gateway/service/getThings.go:101]: Doesn't create a `string` but `uri.URI`

File: http-gateway/service/getResource_test.go [hunk 5]
```diff
@@ -215,3 +216,87 @@ func TestRequestHandlerGetResource(t *testing.T) {
...
+
+	shutdownHttp := httpgwTest.SetUp(t)
+	defer shutdownHttp()
+
+	token := oauthTest.GetDefaultAccessToken(t)
+	ctx = kitNetGrpc.CtxWithToken(ctx, token)
+
+	conn, err := grpc.NewClient(config.GRPC_GW_HOST, grpc.WithTransportCredentials(credentials.NewTLS(&tls.Config{
+		RootCAs: test.GetRootCertificatePool(t),
+	})))
```
Comment (Danielius1922) [inline http-gateway/service/getResource_test.go:237]: Let's ignore it for tests, we're testing other functionality

File: http-gateway/service/getResource_test.go [hunk 6]
```diff
@@ -215,3 +216,87 @@ func TestRequestHandlerGetResource(t *testing.T) {
...
+	}()
+	c := pb.NewGrpcGatewayClient(conn)
+
+	_, shutdownDevSim := test.OnboardDevSim(ctx, t, c, deviceID, config.ACTIVE_COAP_SCHEME+"://"+config.COAP_GW_HOST, test.GetAllBackendResourceLinks())
+	defer shutdownDevSim()
+
+	type args struct {
+		deviceID          string
+		resourceHref      string
+		resourceInterface string
```
Comment (Danielius1922) [inline http-gateway/service/getResource_test.go]: only the first two members are used (`deviceID` and `resourceHref`), others should be removed

File: http-gateway/service/updateResource_test.go [hunk 7]
```diff
@@ -229,3 +232,107 @@ func TestRequestHandlerUpdateResourcesValues(t *testing.T) {
...
+	deviceID := test.MustFindDeviceByName(test.TestDeviceName)
+	switchID := "1"
+	type args struct {
+		accept            string
+		contentType       string
+		deviceID          string
+		resourceHref      string
+		resourceInterface string
+		resourceData      interface{}
+		ttl               time.Duration
```
Comment (Danielius1922) [inline http-gateway/service/updateResource_test.go]: `resourceInterface ` and `ttl` are unused

File: http-gateway/service/getResource_test.go (hunk 6 above)
Comment (jkralik) [reply http-gateway/service/getResource_test.go]: removed

File: http-gateway/service/updateResource_test.go (hunk 7 above)
Comment (jkralik) [reply http-gateway/service/updateResource_test.go]: removed

**Review by jkralik — APPROVED** [review]

**Pull request merged**